core/shared/src/main/scala/zio/ZManaged.scala

```diff
@@ -934,49 +904,36 @@ final class ZManaged[-R, +E, +A] private (reservation: ZIO[R, E, Reservation[R,
    * If the reservation completes successfully (even after the timeout) the release action will be run on a new fiber.
    * `Some` will be returned if acquisition and reservation complete in time
    */
-  def timeout(d: Duration): ZManaged[R with Clock, E, Option[A]] = {
-    def timeoutReservation[B](
-      zio: ZIO[R, E, Reservation[R, E, A]],
-      d: Duration
-    ): ZIO[R with Clock, E, Option[(Duration, Reservation[R, E, A])]] =
-      clock.nanoTime.flatMap { start =>
-        zio
-          .raceWith(ZIO.sleep(d).interruptible)(
-            {
-              case (leftDone, rightFiber) =>
-                rightFiber.interrupt.flatMap(_ =>
-                  leftDone.foldM(
-                    ZIO.halt(_),
-                    succ => clock.nanoTime.map(end => Some((Duration.fromNanos(end - start), succ)))
-                  )
-                )
-            }, {
-              case (exit, leftFiber) =>
-                val cleanup = leftFiber.await
-                  .flatMap(
-                    _.foldM(
-                      _ => ZIO.unit,
-                      _.release(exit)
-                    )
-                  )
-                  .uninterruptible
-                cleanup.fork.as(None).uninterruptible
-            }
-          )
-      }
-
+  def timeout(d: Duration): ZManaged[R with Clock, E, Option[A]] =
     ZManaged {
-      timeoutReservation(reserve, d).map {
-        case Some((spentTime, Reservation(acquire, release))) if spentTime < d =>
-          Reservation(acquire.timeout(Duration.fromNanos(d.toNanos - spentTime.toNanos)), release)
-        case Some((_, Reservation(_, release))) =>
-          Reservation(ZIO.succeedNow(None), release)
-        case _ => Reservation(ZIO.succeedNow(None), _ => ZIO.unit)
+      ZIO.uninterruptibleMask { restore =>
+        for {
+          env                  <- ZIO.environment[(R with Clock, ReleaseMap)]
+          (r, outerReleaseMap) = env
+          innerReleaseMap      <- ZManaged.ReleaseMap.make
+          earlyRelease         <- outerReleaseMap.add(innerReleaseMap.releaseAll(_, ExecutionStrategy.Sequential))
+          raceResult <- restore {
+                         zio
+                           .provide((r, innerReleaseMap))
+                           .raceWith(ZIO.sleep(d).as(None))(
+                             (result, sleeper) => sleeper.interrupt *> ZIO.done(result.map(tp => Right(tp._2))),
+                             (_, resultFiber) => UIO.succeed(Left(resultFiber))
+                           )
+                           .provide(r)
+                       }
+          a <- raceResult match {
+                case Right(value) => UIO.succeed(Some(value))
+                case Left(fiber) =>
+                  ZIO.fiberId.flatMap { id =>
+                    fiber.interrupt
```

Comment:
I think you could pass id here too right?

Reply:
`Fiber#interrupt` automatically uses the current fiber's ID as the interruptor.